homeassistant/components/recorder/__init__.py

```diff
@@ -367,15 +369,13 @@ def _purge_events(session):
         if self._commit(_purge_events):
             _LOGGER.info("Purged events created before %s", purge_before)
 
-        Session.expire_all()
-
         # Execute sqlite vacuum command to free up space on disk
         if self.engine.driver == 'sqlite':
             _LOGGER.info("Vacuuming SQLite to free space")
             self.engine.execute("VACUUM")
 
     @staticmethod
-    def _commit(work):
+    def _commit(work, close_session: Optional[bool]=True):
         """Commit & retry work: Either a model or in a function."""
         import sqlalchemy.exc
         session = Session()
```

Comment:
What about this assignment of a new session? Wouldn't this mess up everything? Should this be removed? 

----

Actually, I think that this PR is causing a lot of confusion. We should move closing the session to the responsibility of the code that is initiating the query. Let's not be smart with trying to auto close sessions.

So add a Session().close() to `run_information` but remove it from `execute` and `_commit`.

http://docs.sqlalchemy.org/en/latest/orm/session_basics.html#when-do-i-construct-a-session-when-do-i-commit-it-and-when-do-i-close-it

> For a GUI interface-driven application, the scope of the Session may best be within the scope of a user-generated event, such as a button push. Or, the scope may correspond to explicit user interaction, such as the user “opening” a series of records, then “saving” them.

They actually have a great pattern that we should use, a context manager. From the same page as linked above:

```python
### another way (but again *not the only way*) to do it ###

from contextlib import contextmanager

@contextmanager
def session_scope():
    """Provide a transactional scope around a series of operations."""
    session = Session()
    try:
        yield session
        session.commit()
    except:
        session.rollback()
        raise
    finally:
        session.close()


def run_my_program():
    with session_scope() as session:
        ThingOne().go(session)
        ThingTwo().go(session)
```